lib-ee/emqx_ee_connector/src/emqx_ee_connector_clickhouse.erl

```diff
@@ -0,0 +1,444 @@
+    }),
+    PoolName = emqx_plugin_libs_pool:pool_name(InstanceID),
+    Options = [
+        {url, URL},
+        {user, maps:get(username, Config, "default")},
+        {key, emqx_secret:wrap(maps:get(password, Config, "public"))},
+        {database, DB},
+        {auto_reconnect, ?AUTO_RECONNECT_INTERVAL},
+        {pool_size, PoolSize},
+        {pool, PoolName}
+    ],
+    InitState = #{poolname => PoolName},
+    try
+        Templates = prepare_sql_templates(Config),
+        State = maps:merge(InitState, #{templates => Templates}),
+        case emqx_plugin_libs_pool:start_pool(PoolName, ?MODULE, Options) of
+            ok ->
+                {ok, State};
+            {error, Reason} ->
+                log_start_error(Config, Reason, none),
+                {error, Reason}
+        end
+    catch
+        _:CatchReason:Stacktrace ->
+            log_start_error(Config, CatchReason, Stacktrace),
+            {error, CatchReason}
+    end.
+
+log_start_error(Config, Reason, Stacktrace) ->
+    StacktraceMap =
+        case Stacktrace of
+            none -> #{};
+            _ -> #{stacktrace => Stacktrace}
+        end,
+    LogMessage =
+        #{
+            msg => "clickhouse_connector_start_failed",
+            error_reason => Reason,
+            config => emqx_misc:redact(Config)
+        },
+    ?SLOG(info, maps:merge(LogMessage, StacktraceMap)),
+    ?tp(
+        clickhouse_connector_start_failed,
+        #{error => Reason}
+    ).
+
+%% Helper functions to prepare SQL tempaltes
+
+prepare_sql_templates(#{
+    sql := Template,
+    batch_value_separator := Separator
+}) ->
+    InsertTemplate =
+        emqx_plugin_libs_rule:preproc_tmpl(Template),
+    BulkExtendInsertTemplate =
+        prepare_sql_bulk_extend_template(Template, Separator),
+    #{
+        send_message_template => InsertTemplate,
+        extend_send_message_template => BulkExtendInsertTemplate
+    };
+prepare_sql_templates(_) ->
+    %% We don't create any templates if this is a non-bridge connector
+    #{}.
+
+prepare_sql_bulk_extend_template(Template, Separator) ->
+    ValuesTemplate = split_clickhouse_insert_sql(Template),
+    %% The value part has been extracted
+    %% Add separator before ValuesTemplate so that one can append it
+    %% to an insert template
+    ExtendParamTemplate = iolist_to_binary([Separator, ValuesTemplate]),
+    emqx_plugin_libs_rule:preproc_tmpl(ExtendParamTemplate).
+
+%% This function is similar to emqx_plugin_libs_rule:split_insert_sql/1 but can
+%% also handle Clickhouse's SQL extension for INSERT statments that allows the
+%% user to specify different formats:
+%%
+%% https://clickhouse.com/docs/en/sql-reference/statements/insert-into/
+%%
+split_clickhouse_insert_sql(SQL) ->
+    ErrorMsg = <<"The SQL template should be an SQL INSERT statement but it is something else.">>,
+    case
+        re:split(SQL, "(\\s+(?i:values)|(?i:format\\s+(?:[A-Za-z0-9_])+)\\s+)", [{return, binary}])
```

Comment:
I noticed that we also have `emqx_plugin_libs_rule:split_insert_sql`, which seems to only cover the first part of this regex more or less.  Should we improve that version and use that here?

Reply:
The format syntax that clickhouse has is not standard SQL so I don't think that will be useful for the other databases. Therefore, I think it is fine to have a specialized version just for clickhouse. What do you think?

Reply:
Sure, I just wanted to bring that other function to attention so we'd avoid missing fixes between the bridges that use SQL, in case both used the same logic and a bug were to be found affecting both.